firesql/sql/sql_fire_query.py

```diff
@@ -250,7 +253,10 @@ def _get_field_value(self, doc, field):
 
   def _get_join_part(self, documents: Dict, partRef: SQL_SelectFrom, isStar: bool = False):
     part, field = partRef
-    docs = documents[part]
+    try:
+      docs = documents[part]
+    except Exception as e:
+      raise Exception(f"error retrieving documents from collection '{part}'")
```

Comment:
This is part of the major JOINS fix.  Previously, executing a JOIN required a WHERE clause.  I think there is more work to be done here because one still cannot write a SQL JOIN without specifying a WHERE clause.  The use-case in doing so is to retrieve all documents matching the JION.  For instance, if one does not mention both tables in the WHERE clause of a corresponding JOIN, this would resulting in the key not being found.  The exception thrown in this case - believe it or not - contained ONLY the literal key name (and no more) in the exception message.  So if one does not include one of the required mentions of the tables involved in the join, then at least the exception message is a bit more descriptive.

Reply:
Also, with the addition of the "!=" operator to joins, here is an example on how to select ALL documents where the JOIN condition is true:

```
 SELECT 
    ds_ids.docid, ds_ids.ds_name, ds_users.docid, ds_users.username, ds_users.created, ds_users.updated
FROM 
    ds_ids JOIN ds_users 
    ON ds_ids.docid = ds_users.ds_id
WHERE 
    ds_ids.docid IS NOT NULL
    AND ds_users.ds_id IS NOT NULL
```
Note that the inclusion of both tables in the WHERE clause is still required - only including one or none will still throw an exception.

Also note that this SQL example uses virtually every "fix" in the PR.

Reply:
This is undoubtedly a design flaw in my original JOIN implementation assuming the data must be retrieved before operating on the JOIN. I am happy to see the improvements in the JOIN logic that you supplied. Thanks